### PR TITLE
8299505: findVirtual on array classes incorrectly restricts the receiver type

### PR DESCRIPTION
**Pull request by liach**
The access hack for array class clone is only applied to `checkAccess` but missing before call to `restrictProtectedReceiver`, causing the array receiver type to be incorrectly replaced by the lookupClass type. This patch fixes that and adds a test to ensure an original lookup resolves `clone` for both array classes (public) and Object (inherited protected) correctly, and restores the old MethodHandlesGeneralTest from [JDK-8001105](https://bugs.openjdk.org/browse/JDK-8001105) which ensures correctness for publicLookup (which is already correct).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299505](https://bugs.openjdk.org/browse/JDK-8299505): findVirtual on array classes incorrectly restricts the receiver type

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [532ea964](https://git.openjdk.org/jdk/pull/13855/files/532ea964d6777131105aa56742576d9d28f13bc2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13855/head:pull/13855` \
`$ git checkout pull/13855`

Update a local copy of the PR: \
`$ git checkout pull/13855` \
`$ git pull https://git.openjdk.org/jdk.git pull/13855/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13855`

View PR using the GUI difftool: \
`$ git pr show -t 13855`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13855.diff">https://git.openjdk.org/jdk/pull/13855.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13855#issuecomment-1537196696)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -4016,6 +4020,8 @@ private void checkSpecialCaller(Class<?> specialCaller, Class<?> refc) throws Il
         private boolean restrictProtectedReceiver(MemberName method) {
             // The accessing class only has the right to use a protected member
             // on itself or a subclass.  Enforce that restriction, from JVMS 5.4.4, etc.
+            // Since array have access hacks, need to guard against array clone() before
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:4023]: I suggest to move this new comment to `getDirectMethodCommon`, the relevant caller of this method.  See suggested wordings below.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -4101,7 +4107,9 @@ private MethodHandle getDirectMethodCommon(byte refKind, Class<?> refc, MemberNa
             MethodHandle mh = dmh;
             // Optionally narrow the receiver argument to lookupClass using restrictReceiver.
             if ((doRestrict && refKind == REF_invokeSpecial) ||
-                    (MethodHandleNatives.refKindHasReceiver(refKind) && restrictProtectedReceiver(method))) {
+                    (MethodHandleNatives.refKindHasReceiver(refKind) &&
+                            !isArrayClone(refKind, refc, method) &&
+                            restrictProtectedReceiver(method))) {
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:4112]:
```suggestion
                    (MethodHandleNatives.refKindHasReceiver(refKind) &&
                            restrictProtectedReceiver(method) &&
                            // All arrays simply inherit the protected Object.clone method.
                            // The leading argument is restricted to the requested array type
                            // (not the lookup class).
                            !isArrayClone(refKind, refc, method))) {
```

File: test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8299505
+ * @run junit FindVirtualArrayCloneTest
+ * @summary Ensures Arrays' clone doesn't have incorrect receiver type bound
```
Comment (mlchung) [inline test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java:28]:
```suggestion
 * @summary Test invocation of Object.clone for arrays
```

Comment (mlchung) [reply test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java:28]: Can you also verify that the invocation of `Object.clone` returns an array object with correct content for completeness?

**Review by mlchung — COMMENTED** [review]
Thanks for taking this.   The patch looks right.    It should special case `Object.clone` for array types not to restrict the receiver to the lookup class.

File: test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java (hunk 3 above)
Comment (liach) [reply test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java]: I've done the requested changes and ran the invoke tests locally.

**Review by mlchung — APPROVED** [review]
Looks good.

File: test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        assertSame(lookup.lookupClass(), cloneMh.type().parameterType(0));
+    }
+
+    @Test
+    public void checkArrayClone() throws Throwable {
+        var array = new String[] {"apple", "pear", "banana"};
+        var regularClone = array.clone();
+        var cloneMh = MethodHandles.lookup()
+                .findVirtual(String[].class, "clone", MethodType.methodType(Object.class));
+        var mhClone = (String[]) (Object) cloneMh.invokeExact((String[]) array);
```
[the remark this answers is not in the document]
Comment (liach) [reply test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java:70]: the clone Method handle will have a type of `(String[])Object`, as `findVirtual` for `clone` uses the Object return method type, and there is no covariant override in array classes (which don't declare any methods)

[the remark this answers is not in the document]
Comment (mlchung) [reply test/jdk/java/lang/invoke/findVirtual/FindVirtualArrayCloneTest.java:70]: You're right.  Alternatively (perhaps easier to read):

```
-                .findVirtual(String[].class, "clone", MethodType.methodType(Object.class));
-        var mhClone = (String[]) (Object) cloneMh.invokeExact((String[]) array);
+                .findVirtual(String[].class, "clone", MethodType.methodType(Object.class))
+                .asType(MethodType.methodType(String[].class, String[].class));
+        var mhClone = (String[]) cloneMh.invokeExact((String[]) array);
```

**Pull request closed without merging**